File: src/main/java/com/zynchub/digital/hubservice/app/controller/AuthController.java
package com.zynchub.digital.hubservice.app.controller;

import com.zynchub.digital.hubservice.app.config.AuthCognitoProperties;
import jakarta.servlet.http.HttpServletRequest;
import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.UUID;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.util.StringUtils;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/auth")
public class AuthController {

  private final AuthCognitoProperties.Properties properties;

  public AuthController(AuthCognitoProperties.Properties properties) {
    this.properties = properties;
  }

  @GetMapping("/login")
  public ResponseEntity<Void> login() {
    if (!properties.enabled()) {
      return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
    }

    String state = UUID.randomUUID().toString();
    String scope = String.join(" ", defaultScopes(properties.scopes()));

    String authorizeUrl =
        String.format(
            "%s/oauth2/authorize?client_id=%s&response_type=code&scope=%s&redirect_uri=%s&state=%s",
            properties.domain(),
            urlEncode(properties.clientId()),
            urlEncode(scope),
            urlEncode(properties.redirectUri()),
            urlEncode(state));

    HttpHeaders headers = new HttpHeaders();
    headers.setLocation(URI.create(authorizeUrl));
    return new ResponseEntity<>(headers, HttpStatus.FOUND);
  }

  @GetMapping("/callback")
  public ResponseEntity<Void> callback(
      @RequestParam(name = "code", required = false) String code,
      @RequestParam(name = "state", required = false) String state,
      HttpServletRequest request) {
    if (!properties.enabled()) {
      return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
    }

    if (!StringUtils.hasText(code)) {
      return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
    }

    String redirect = properties.postLoginRedirectUri();
    String target =
        String.format(
            "%s?code=%s%s",
            redirect,
            urlEncode(code),
            StringUtils.hasText(state) ? "&state=" + urlEncode(state) : "");

    HttpHeaders headers = new HttpHeaders();
    headers.setLocation(URI.create(target));
    return new ResponseEntity<>(headers, HttpStatus.FOUND);
  }

  private static List<String> defaultScopes(List<String> scopes) {
    return (scopes == null || scopes.isEmpty()) ? List.of("openid", "email", "profile") : scopes;
  }

  private static String urlEncode(String value) {
    return URLEncoder.encode(value, StandardCharsets.UTF_8);
  }
}

File: internal/app/controller/auth_controller.go
package controller

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"

	"plexus-bff-service-go/internal/app/config"
)

type AuthController struct {
	properties config.CognitoConfig
}

func NewAuthController(properties config.CognitoConfig) *AuthController {
	return &AuthController{properties: properties}
}

func (c *AuthController) Login(w http.ResponseWriter, r *http.Request) {
	if !c.properties.Enabled {
		w.WriteHeader(http.StatusNotFound)
		return
	}

	scopes := c.properties.Scopes
	if scopes == "" {
		scopes = "openid,email,profile"
	}
	target := fmt.Sprintf("%s/oauth2/authorize?client_id=%s&response_type=code&scope=%s&redirect_uri=%s&state=%s",
		c.properties.Domain,
		url.QueryEscape(c.properties.ClientID),
		url.QueryEscape(strings.Join(splitScopes(scopes), " ")),
		url.QueryEscape(c.properties.RedirectURI),
		url.QueryEscape("generated-state"),
	)
	http.Redirect(w, r, target, http.StatusFound)
}

func (c *AuthController) Callback(w http.ResponseWriter, r *http.Request) {
	if !c.properties.Enabled {
		w.WriteHeader(http.StatusNotFound)
		return
	}
	code := r.URL.Query().Get("code")
	state := r.URL.Query().Get("state")
	if code == "" {
		w.WriteHeader(http.StatusBadRequest)
		return
	}
	target := fmt.Sprintf("%s?code=%s", c.properties.PostLoginRedirectURI, url.QueryEscape(code))
	if state != "" {
		target += "&state=" + url.QueryEscape(state)
	}
	http.Redirect(w, r, target, http.StatusFound)
}

func splitScopes(scopes string) []string {
	parts := strings.Split(scopes, ",")
	result := make([]string, 0, len(parts))
	for _, part := range parts {
		part = strings.TrimSpace(part)
		if part != "" {
			result = append(result, part)
		}
	}
	if len(result) == 0 {
		return []string{"openid", "email", "profile"}
	}
	return result
}
